### PR TITLE
Tools: Add installer translation checker

### PR DESCRIPTION
**Pull request by hoffie**
Compares Windows installer translations with the English version, checks for consistency and common mistakes.

This is helpful for reviewing installer translation PRs. We may consider putting it into CI runs at some point. Not suggesting this right now though.

### CODE REVIEW

File: tools/check-wininstaller-translations.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+set -eu
+
+BASE_DIR=src/res/translation/wininstaller/
+BASE_LANG=en
+INSTALLERLNG=installerlng.nsi
+BASE_LANGSTRINGS=$(grep LangString "${BASE_DIR}/${BASE_LANG}.nsi" | cut -d' ' -f2)
+EXIT=0
+for LANGUAGE_FILE in $(ls -1 src/res/translation/wininstaller/??.nsi | grep -vF "${BASE_LANG}.nsi"); do
+		echo
```
Comment (pljones) [inline tools/check-wininstaller-translations.sh:10]: What are you using for indent?  It's huge!

File: tools/check-wininstaller-translations.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		echo "* ${LANGUAGE_FILE}"
+		echo -n "  - Checking language file is included in ${INSTALLERLNG}... "
+		if grep -qF $(tr '/' '\\' <<<"${LANGUAGE_FILE}") "${BASE_DIR}/${INSTALLERLNG}"; then
+				echo "ok"
```
Comment (pljones) [inline tools/check-wininstaller-translations.sh:14]: Two tabs, I see...  That's rather a lot for a single level indent.

File: tools/check-wininstaller-translations.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		echo -n "  - Checking for multiple lang macros... "
+		if [[ $(sort -u <<<"$LANG_MACROS" | wc -l) != 1 ]]; then
+				echo "ERROR, found multiple LANG_ macros"
+				EXIT=1
+		else
+				echo "ok"
+		fi
+
+		echo -n "  - Checking if LANG_ macro is in ${INSTALLERLNG}..."
+		if grep -qi '!insertmacro MUI_LANGUAGE "'$(sort -u <<<"${LANG_MACROS}" | sed -rne 's/\$\{LANG_(.*)\}/\1/p')'"' "$BASE_DIR/${INSTALLERLNG}"; then
```
Comment (pljones) [inline tools/check-wininstaller-translations.sh:54]: You could split then `; then` onto the next line - it looks nicer in bash, I think.

File: tools/check-wininstaller-translations.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+				EXIT=1
+		fi
+
+		echo -n "  - Checking for LangStrings diffs... "
+		OTHER_LANGSTRINGS=$(grep LangString "${LANGUAGE_FILE}" | cut -d' ' -f2)
+		diff=$(diff -u <(echo "$BASE_LANGSTRINGS") <(echo "$OTHER_LANGSTRINGS"))
+		[[ ! "$diff" ]] && echo "ok" && continue
+		echo "warning, differences found:"
+		echo "---"
+		echo "$diff"
```
Comment (pljones) [inline tools/check-wininstaller-translations.sh:67]: How does the formatting from the diff look here (i.e. are all the line breaks still there)?  If it's lost, just rerun the diff.

File: tools/check-wininstaller-translations.sh [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		echo "---"
+		echo
+		EXIT=1
```
Comment (pljones) [inline tools/check-wininstaller-translations.sh:70]: I'd do this immediately after the `if` on 64.

Comment (hoffie) [reply tools/check-wininstaller-translations.sh]: Yep, done. Thanks for all your suggestions!

File: tools/check-wininstaller-translations.sh (hunk 4 above)
Comment (hoffie) [reply tools/check-wininstaller-translations.sh]: Line breaks are still there due to proper quoting :)

```
* src/res/translation/wininstaller/de.nsi
  - Checking language file is included in installerlng.nsi... ok
  - Checking language file for empty strings... ok
  - Checking for comment leftovers from original file... ok
  - Checking for wrong macros... ok
  - Checking for multiple lang macros... ok
  - Checking if LANG_ macro is in installerlng.nsi...ok
  - Checking for LangStrings diffs... warning, differences found:
--- /dev/fd/63	2021-02-28 10:06:27.394345680 +0100
+++ /dev/fd/62	2021-02-28 10:06:27.397679038 +0100
@@ -2 +2 @@
-INVALID_FOLDER_MSG
+INVALID_FOLDER_MSGx
```

File: tools/check-wininstaller-translations.sh (hunk 3 above)
Comment (hoffie) [reply tools/check-wininstaller-translations.sh]: > You could split then `; then` onto the next line - it looks nicer in bash, I think, and means the`;` is not required.

Hrm, I'd like to keep the `then` on the same line as other scripts do it the same way (e.g. `./linux/autorelease_linux.sh`, `./android/automated_build/install-qt.sh`). I think both styles are common though.

> (You could also split the long `$(...)` over a couple of lines with line breaks after and before the `(` and `)`, plus a break with `|\` before the `sed`. i.e. one command per line.)

I've moved the long call to a variable which makes the line shorter.

File: tools/check-wininstaller-translations.sh (hunk 1 above)
Comment (hoffie) [reply tools/check-wininstaller-translations.sh]: Hm, no idea where I've got those defaults from. Changed to 4 spaces now to match `./linux/autorelease_linux.sh`.

File: tools/check-wininstaller-translations.sh (hunk 2 above)
Comment (hoffie) [reply tools/check-wininstaller-translations.sh]: See above.

File: tools/check-wininstaller-translations.sh [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    if grep -qF $(tr '/' '\\' <<<"${LANGUAGE_FILE}") "${BASE_DIR}/${INSTALLERLNG}"; then
+        echo "ok"
+    else
+        echo "ERROR"
+        EXIT=1
+    fi
+
+    echo -n "  - Checking language file for empty strings... "
+    if grep -qF '""' "${LANGUAGE_FILE}"; then
+        echo "ERROR, found empty string(s)"
```
Comment (ann0see) [inline tools/check-wininstaller-translations.sh:23]: I'm not sure if we want to output an error here or just a warning?

Comment (hoffie) [reply tools/check-wininstaller-translations.sh]: Ok, changed to warning.

**Review by ann0see — APPROVED** [review]
I think, this can be merged soon.

**Review by pljones — APPROVED** [review]

**Pull request merged**